README
Crypt/OpenSSL/PKCS12 version 1.8
===============================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  OpenSSL Development

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2004-2021 Dan Sully

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension Crypt::OpenSSL::PKCS12.

1.8 Fri Nov 12 20:31:24 CET 2021

- Applied patch via PR #37 from @SmartCodeMaker, calculating of strings and handling Windows files properly

- Applied patch via PR #38 from @SmartCodeMaker, formatting print of SVs displays garbage as documented in the perlguts documentation

  REF: https://perldoc.perl.org/perlguts#Formatted-Printing-of-SVs

1.7  Fri May 21 21:16:53 CEST 2021

- Removed some unnecessary code from a test, the code would not compile on Perl 5.18

1.6  Thu May 20 07:32:35 CEST 2021

- Added encoding declaration to POD section, I had overlooked a special character in one of the names of the contributors

1.5  Wed May 19 22:03:48 CEST 2021

- Addressed minor issue with the implementation added in 1.4, fix lifted from Crypt::OpenSSL::X509 - Thanks Shoichi Kaji

- Added section on contributors to the POD

1.4  Sun Apr 11 16:23:54 CEST 2021

- Addressed issue with homebrew path reported for Crypt::OpenSSL::X509

  REF: https://github.com/dsully/perl-crypt-openssl-x509/issues/81

  The pattern is the same in this distribution

1.3  Thu Jun  4 21:39:30 CEST 2020

- Maintenance release addressing the broken inc/ configuration pointed out in: RT:77784 by HMBRAND

  REF: https://rt.cpan.org/Ticket/Display.html?id=77784

1.2  Sat Oct 27 09:37:04 CEST 2018

- Maintenance release addressing the missing exclusion of the MYMETA files issue spotted by Alexandr Ciornii (@chorny) #issue32

1.1  Fri Oct 26 07:54:48 CEST 2018

- PR from Todd Rinaldo (@toddr) Adding use of our instead of vars, since we are now bumping Perl version requirement
  from 5.005 to 5.6.0

- PR from Todd Rinaldo (@toddr) Exchange DynaLoader for XSLoader, bumping Perl version requirement
  from 5.005 to 5.6.0

- PR from Todd Rinaldo (@toddr) Add Makefile.PL support for perl 5.26.0 and above which don't include . in @INC

1.0  Mon Mar 19 21:20:49 CET 2018

- Re-release of 0.10, due to indexing issue with PAUSE/CPAN ref: #28

0.10  Wed Mar 14 08:11:54 CET 2018

- Patch from kelson (@kelson42) improving the code, going from two pass to one pass, speeding up the chain management PR #27

0.9   Sun Mar 11 19:45:18 CET 2018

- Patch from kelson (@kelson42) addressing compilation issues against OpenSSL 1.1.0 PR #24

0.8   Sun Nov 20 19:21:47 CET 2017

- Patch from jonasbn improving handling of C include paths

- Patch from jonasbn reinitializing inc/ and Module::Install integration

- Patch from Songmu adjusting use of paths and flags in Makefile.PL

- Patch from jonasbn adding inc/Module/Install/External.pm from Module::Install missing from MANIFEST

- Patch from jonasbn for compatibility with required C libraries installed via Homebrew on OSX

- Patch from jonasbn to address clang changes

- Patch from Christopher Hoskin enabling compatibility with OpenSSL 1.1.0 and backwards compatibility with OpenSSL 1.0.2

0.7   Fri Oct 24 06:16:43 PDT 2014

- Patch from Mikołaj Zalewski to fix C90 issue on CentOS.

- Patch from Darko Prelec to fix compile issue #2.

0.6   Thu May 10 10:44:30 PDT 2012

- Patch from Leonid Antonenkov to implement private key access.

- Misc compile fixes.

0.5   Sun May 18 08:19:14 PDT 2008

- Fix pod coverage tests.

0.4   Sat May 17 00:47:36 PDT 2008

- Stop cpansmoke from emailing me.

0.3   Fri Jan  5 15:36:28 PST 2007

- Add more POD & coverage.
- Fix META.yml

0.2   Thu Jan  4 10:21:31 PST 2007

- Use Module::Install

- Added tests.

- PKCS12 Create from JoNO

0.1   Thu Jan 29 17:01:38 PST 2004

- Initial release.


test.pl

use Test;

BEGIN { plan tests => 1 };

use Crypt::OpenSSL::PKCS12;

ok(1);



PKCS12.pm
package Crypt::OpenSSL::PKCS12;

use warnings;
use strict;
use Exporter;

our $VERSION = '1.8';
our @ISA = qw(Exporter);

our @EXPORT_OK = qw(NOKEYS NOCERTS INFO CLCERTS CACERTS);

use XSLoader;

XSLoader::load 'Crypt::OpenSSL::PKCS12', $VERSION;

END {
  __PACKAGE__->__PKCS12_cleanup();
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Crypt::OpenSSL::PKCS12 - Perl extension to OpenSSL's PKCS12 API.

=head1 SYNOPSIS

  use Crypt::OpenSSL::PKCS12;

  my $pass   = "your password";
  my $pkcs12 = Crypt::OpenSSL::PKCS12->new_from_file('cert.p12');

  print $pkcs12->certificate($pass);
  print $pkcs12->private_key($pass);

  if ($pkcs12->mac_ok($pass)) {
  ...

  # Creating a file
  $pkcs12->create('test-cert.pem', 'test-key.pem', $pass, 'out.p12', 'friendly name');


  # Creating a string
  my $pksc12_data = $pkcs12->create_as_string(
    catdir($base, 'test-cert.pem'),
    catdir($base, 'test-key.pem'),
    $pass,
    'Friendly Name'
  );

=head1 DESCRIPTION

This implements a small bit of OpenSSL's PKCS12 API.

=head1 FUNCTIONS

=over 4

=item * new( )

=item * new_from_string( C<$string> )

=item * new_from_file( C<$filename> )

Create a new Crypt::OpenSSL::PKCS12 instance.

=item * certificate( [C<$pass>] )

Get the Base64 representation of the certificate.

=item * private_key( [C<$pass>] )

Get the Base64 representation of the private key.

=item * as_string( [C<$pass>] )

Get the binary represenation as a string.

=item * mac_ok( [C<$pass>] )

Verifiy the certificates Message Authentication Code

=item * changepass( C<$old>, C<$new> )

Change a certificate's password.

=item * create( C<$cert>, C<$key>, C<$pass>, C<$output_file>, C<$friendly_name> )

Create a new PKCS12 certificate. $cert & $key may either be strings or filenames.

C<$friendly_name> is optional.

=item * create_as_string( C<$cert>, C<$key>, C<$pass>, C<$friendly_name> )

Create a new PKCS12 certificate string. $cert & $key may either be strings or filenames.

C<$friendly_name> is optional.

Returns a string holding the PKCS12 certicate.

=back

=head1 EXPORT

None by default.

On request:

=over 4

=item * C<NOKEYS>

=item * C<NOCERTS>

=item * C<INFO>

=item * C<CLCERTS>

=item * C<CACERTS>

=back

=head1 SEE ALSO

=over

=item * OpenSSL(1)

=item * L<Crypt::OpenSSL::X509|https://metacpan.org/pod/Crypt::OpenSSL::X509>

=item * L<Crypt::OpenSSL::RSA|https://metacpan.org/pod/Crypt::OpenSSL::RSA>

=item * L<Crypt::OpenSSL::Bignum|https://metacpan.org/pod/Crypt::OpenSSL::Bignum>

=back

=head1 AUTHOR

=over

=item * Dan Sully, E<lt>daniel@cpan.orgE<gt>

=back

Current maintainer

=over

=item * jonasbn

=back

=head1 CONTRIBUTORS

In alphabetical order, contributors, bug reporters and all

=over

=item * @mmuehlenhoff

=item * @sectokia

=item * @SmartCodeMaker

=item * Alexandr Ciornii, @chorny

=item * Christopher Hoskin, @mans0954

=item * Daisuke Murase, @typester

=item * Darko Prelec, @dprelec

=item * David Steinbrunner, @dsteinbrunner

=item * Giuseppe Di Terlizzi, @giterlizzi

=item * H.Merijn Brand, @tux

=item * Hakim, @osfameron

=item * J. Nick Koston, @bdraco

=item * James Rouzier, @jrouzierinverse

=item * jonasbn. @jonasbn

=item * Kelson, @kelson42

=item * Lance Wicks, @lancew

=item * Leonid Antonenkov

=item * Masayuki Matsuki, @songmu

=item * Mikołaj Zalewski

=item * Shoichi Kaji

=item * Slaven Rezić

=item * Todd Rinaldo, @toddr

=back

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2021 by Dan Sully

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut


inc/MyMakeMaker.pm
package inc::MyMakeMaker;

use Moose;
use File::Spec;

extends 'Dist::Zilla::Plugin::MakeMaker::Awesome';

override _build_WriteMakefile_args => sub {
    my $inc;
    my $libs;
    my $ccflags;

    if ($^O ne 'MSWin32' and my $prefix = `brew --prefix openssl 2>@{[File::Spec->devnull]}`) {
        chomp $prefix;

        $inc  = "-I$prefix/include";
        $libs = "-L$prefix/lib -lcrypto -lssl";
    } else {
        $inc = '-I/usr/local/opt/openssl/include -I/usr/local/include/openssl -I/usr/include/openssl -I/usr/local/include/ssl -I/usr/local/ssl/include';
        $libs = '-L/usr/local/opt/openssl/lib -L/usr/local/lib -L/usr/lib -L/usr/local/ssl/lib -lcrypto -lssl';
    }

    if ($^O =~ m/darwin/i) {
        $ccflags = '-O2 -g -Wall -Werror -Wno-deprecated-declarations -Wno-deprecated-declarations';
     } else {
        $ccflags = '-O2 -g -Wall -Werror';
     }

    return +{
        # Add LIBS => to WriteMakefile() args
        %{ super() },
        LIBS    => $libs,
        INC     => $inc,
        CCFLAGS => $ccflags,
    }
};

__PACKAGE__->meta->make_immutable;


Makefile.PL
# $Id: Makefile.PL 10 1998-12-16 23:02:45Z daniel $

use lib '.';
use inc::Module::Install;

use File::Spec;
use Config;

name('Crypt-OpenSSL-PKCS12');
license('perl');
perl_version('5.006');
all_from('PKCS12.pm');

homepage 'https://github.com/dsully/perl-crypt-openssl-pkcs12';
repository 'https://github.com/dsully/perl-crypt-openssl-pkcs12';
bugtracker 'https://github.com/dsully/perl-crypt-openssl-pkcs12/issues';

test_requires 'Test::Pod::Coverage' => '1.08';
test_requires 'Pod::Coverage' => '0.19';
test_requires 'Test::More' => '0.47';
test_requires 'File::Spec::Functions' => '0'; # in core

requires_external_cc();

if ($^O ne 'MSWin32' and my $prefix = `brew --prefix --installed openssl\@1.1 2>@{[File::Spec->devnull]}`) {
  chomp $prefix;
  inc "-I$prefix/include";
  libs "-L$prefix/lib -lcrypto -lssl";
} else {
  inc '-I/usr/local/opt/openssl@1.1/include -I/usr/local/include/openssl -I/usr/include/openssl -I/usr/local/include/ssl -I/usr/local/ssl/include';
  libs '-L/usr/local/opt/openssl@1.1/lib -L/usr/local/lib -L/usr/lib -L/usr/local/ssl/lib -lcrypto -lssl';
}

my $cc_option_flags = '-O2 -g -Wall -Werror';

if ($Config::Config{gccversion} =~ /llvm/i) {
  $cc_option_flags .= ' -Wno-deprecated-declarations'; # -Wno-incompatible-pointer-types-discards-qualifiers -Wno-compound-token-split-by-macro
  if ($Config::Config{perl_version} <= 20) {
    $cc_option_flags .= ' -Wno-error=pointer-bool-conversion';
  }
}

auto_install();
WriteAll();
